File: test/parser_parity.py
import json
from pathlib import Path
import sys
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from todo_tasks import parse_tasks, set_checked, set_task_text, remove_task, append_task

fixtures = json.load(sys.stdin)
for index, fixture in enumerate(fixtures):
    text = fixture['text']
    tasks = parse_tasks(text)
    line = tasks[0].line if tasks else -1
    added = append_task(text, 'new task')
    result = dict(tasks=[dict(line=t.line, indent=t.indent, checked=t.checked, text=t.text) for t in tasks],
        checked=set_checked(text, line, True), edited=set_task_text(text, line, 'edited'),
        removed=remove_task(text, line), added=dict(text=added[0], line=added[1]) if added else None)
    assert result == fixture['result'], (index, result, fixture['result'])
print(f'UI/scanner parity: {len(fixtures)} markdown fixtures passed')

File: test/parser-parity.js
const fs = require('node:fs');
const src = fs.readFileSync(`${__dirname}/../TodoDoc.js`, 'utf8').replace('.pragma library', '');
const T = new Function(`${src}; return {parse,setChecked,setTaskText,removeTask,appendTask};`)();
const fixtures = [
  '# Title\n\n- [ ] First\n* [X] Done #tag\n',
  '# Title\r\n\r\n- [ ]no gap\r\n  + [X]\ttask\r\n',
  '# Title\r- [ ] mac\r- [ ] final',
  '\uFEFF---\r\ntags: [work]\r\n- [ ] yaml example\r\n---\r\n- [ ] real\r\n',
  '---  \n- [ ] yaml\n... \n- [ ] real\n',
  '````md\n```\n~~~\n- [ ] example\n`````  \n- [ ] real\n',
  '~~~md\n- [ ] example\n~~~\n* [X] real\n',
  '# Notes', '', '\n\n',
  '- [ ] Body\u2028continued\n- [ ] with\vtab\n- [ ] final',
];
console.log(JSON.stringify(fixtures.map(text => {
  const tasks = T.parse(text), line = tasks.length ? tasks[0].line : -1;
  return {text, result:{tasks, checked:T.setChecked(text,line,true), edited:T.setTaskText(text,line,'edited'),
    removed:T.removeTask(text,line), added:T.appendTask(text,'new task')}};
})));
